jina/drivers/multimodal.py

```diff
@@ -0,0 +1,112 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from collections import defaultdict
+from typing import Iterable, Tuple, Dict
+
+import numpy as np
+
+from .encode import BaseEncodeDriver
+from .helper import pb2array, array2pb
+from ..proto import jina_pb2
+
+
+def _extract_doc_content(doc: 'jina_pb2.Document'):
+    """Returns the content of the document with the following priority:
+    If the document has an embedding, return it, otherwise return its content.
+    """
+    if doc.embedding.buffer:
+        return pb2array(doc.embedding)
+    else:
+        return doc.text or doc.buffer or (doc.blob and pb2array(doc.blob))
+
+
+def _extract_modalities_from_document(doc: 'jina_pb2.Document'):
+    """Returns a dictionary of document content (embedding, text, blob or buffer) with `modality` as its key
+    """
+    doc_content = {}
+    for chunk in doc.chunks:
+        modality = chunk.modality
+        if modality in doc_content.keys():
+            return None
+        else:
+            doc_content[modality] = _extract_doc_content(chunk)
+    return doc_content
+
+
+class MultimodalDriver(BaseEncodeDriver):
+    """Extract multimodal embeddings from different modalities.
+    Input-Output ::
+        Input:
+        document:
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+        Output:
+        document: (embedding: multimodal encoding)
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+
+    .. note::
+        - It traverses on the ``documents`` for which we want to apply the ``multimodal`` embedding. This way
+        we can use the `batching` capabilities for the `executor`.
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
```

Comment:
If you think about it, how else can you tell between two modalities by the executor? There must be some difference between every chunk, otherwise they must merge them in previous drivers.